### PR TITLE
chromium: build with upstream llvm & PGO

### PR DESCRIPTION
**Pull request by samuelgrf**
###### Motivation for this change
Google reports up to 10% faster page loads.
Source: https://blog.chromium.org/2020/08/chrome-just-got-faster-with-profile.html

Can decrease overall compile time by 20%.
Source: https://llvm.org/docs/HowToBuildWithPGO.html

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by samuelgrf** [comment]
So far I have only tested that the build starts without errors, I'll report back once it's finished.

**Comment by TredwellGit** [comment]
I am under the impression that https://github.com/NixOS/nixpkgs/pull/101467 already does this.
https://source.chromium.org/chromium/chromium/src/+/master:build/config/compiler/pgo/pgo.gni will set `chrome_pgo_phase = 2` when `is_official_build == true`. https://source.chromium.org/chromium/chromium/src/+/master:build/config/compiler/pgo/BUILD.gn;l=58 downloads PGO profiles.

**Comment by samuelgrf** [comment]
https://source.chromium.org/chromium/chromium/src/+/master:build/config/compiler/pgo/BUILD.gn;l=58 doesn't actually fetch any profile files, it runs the `update_pgo_profiles.py` script with the `get_profile_path` argument, which only gets existing profiles.

To enable fetching profiles https://source.chromium.org/chromium/chromium/src/+/master:DEPS;l=121 can be set to true, which would run the update script at https://source.chromium.org/chromium/chromium/src/+/master:DEPS;l=4961; this approach doesn't work on Nix though, since network access is disabled during the configuration & build phase (to make sure builds are reproducible).

**Comment by TredwellGit** [comment]
Good catch.

**Comment by TredwellGit** [comment]
@primeos

**Comment by samuelgrf** [comment]
Result of `nixpkgs-review pr 101786` [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>chromium</li>
  </ul>
</details>

**Comment by samuelgrf** [comment]
For some reason `nixpkgs-review` only built the stable variant, the binary runs without issues.
I have manually started the beta build and it's running atm.

File: pkgs/applications/networking/browsers/chromium/update.py
Comment (primeos) [inline pkgs/applications/networking/browsers/chromium/update.py:13]: Do we really need to switch to `requests` here (for the `.text` below)? It's not part of the Python standard library so at least we'd have to specify it as a dependency (2nd line, if necessary).

File: pkgs/applications/networking/browsers/chromium/upstream-info.json
Comment (primeos) [inline pkgs/applications/networking/browsers/chromium/upstream-info.json:7]: Would be better to use `pgo-profile{,-sha256}` here (or scope them under a `pgo` key) to keep thinks clearer / easier to read and understand (though I could also change that later).

**Review by primeos — CHANGES_REQUESTED** [review]
Thanks! I have two nitpicks though (at least the 1st one needs to be resolved so that running the script doesn't fail - depending on the installed Python libraries). Other than that it LGTM.

> For some reason nixpkgs-review only built the stable variant, the binary runs without issues.

Could be because we only built `chromium` on Hydra (`meta.hydraPlatforms`), not sure though if `nixpkgs-review` considers that.

**Comment by samuelgrf** [comment]
Weirdly enough every benchmark I've tried has worse results on pgo than on master, I don't think this should be merged until that's fixed.
Does anyone have an idea why that could be?

<details>

Speedometer master: 141.4
Speedometer pgo: 130.5

Kraken master: 696.9ms
Kraken pgo: 716.7ms

Octane master: 54105
Octane pgo: 52454

**Comment by samuelgrf** [comment]
I am pretty sure the performance decrease is caused by mismatching clang versions.

PGO builds are disabled in Ubuntu bionic-dev, the commit message says this was done because `the upstream profile data is not compatible with the version of clang used to build chromium`. [Source](https://bazaar.launchpad.net/~chromium-team/chromium-browser/bionic-dev/revision/1580)
Clang 10 is used in both Ubuntu bionic-dev and nixpkgs. [Source](https://bazaar.launchpad.net/~chromium-team/chromium-browser/bionic-dev/view/head:/debian/control#L20)

Building with mismatched profile data would normally cause compiler errors, but this is disabled in the Chromium build files. [Source](https://source.chromium.org/chromium/chromium/src/+/master:build/config/compiler/pgo/BUILD.gn;l=76)

**Comment by TredwellGit** [comment]
The prebuilt LLVM binaries could be used, but we would need to get [`depot_tools`](https://chromium.googlesource.com/chromium/src/+/master/docs/linux/build_instructions.md#Install) and [`gclient runhooks`](https://chromium.googlesource.com/chromium/src/+/master/docs/linux/build_instructions.md#Run-the-hooks) working. Or better yet, package the [LLVM revisions that Chromium distributes](https://chromium.googlesource.com/chromium/src/+/master/docs/updating_clang.md) that way we don't rely on unverified binaries.

**Comment by samuelgrf** [comment]
AFAIK network access is disabled during Nix builds.
What we could do is add the clang tarball to the update script and fetch it using fetchurl, the relevant data is inside [tools/clang/scripts/update.py](https://source.chromium.org/chromium/chromium/src/+/refs/tags/86.0.4240.111:tools/clang/scripts/update.py;l=41).
The DL link for 86.0.4240.111 would be https://commondatastorage.googleapis.com/chromium-browser-clang/Linux_x64/clang-llvmorg-12-init-3492-ga1caa302-1.tgz .

I just don't know how we'd get the binaries to run, can `autoPatchelfHook` be used for this?

**Comment by TredwellGit** [comment]
Again, it would be better (and maybe easier) to generate `llvmPackages` for each Chromium version then set https://github.com/NixOS/nixpkgs/blob/5eaf0d0d33c9926a98984bb1f0c43f44e0db5100/pkgs/applications/networking/browsers/chromium/default.nix#L25 to those versions.

**Comment by samuelgrf** [comment]
Ok, I've applied the changes mentioned in the review and also implemented upstream llvm builds, nixpkgs-review builds everything without issues (excluding dev, as it's marked as broken).

Benchmark results look better now, [Speedometer](https://browserbench.org/Speedometer2.0/) improved by ~5%, [MS Chalkboard](https://testdrive-archive.azurewebsites.net/Performance/Chalkboard/Default.html) by ~6%, [Octane 2](https://chromium.github.io/octane/) and [Kraken](https://krakenbenchmark.mozilla.org/) by ~0.6%.

I am just not sure how to best handle the upstream llvm builds, I have a few questions (ping @primeos﻿):
1. Should I create a completely new directory in llvmPackages for chromium builds?
With that solution we could remove multiple `if`/`else` statements and simplify the expression, but it'd cause lots of code duplication and the `customVersion` argument might also be useful for other tools like `rocm`.
2. Should I move the llvm expressions to `llvm/12` instead, or should we wait for it's release first?

File: pkgs/applications/networking/browsers/chromium/default.nix
Comment (TredwellGit) [inline pkgs/applications/networking/browsers/chromium/default.nix:33]: I probably should have set `stdenv = llvmPackages.lldClang.stdenv` in https://github.com/NixOS/nixpkgs/pull/101467. Maybe it would be good to include?

**Comment by primeos** [comment]
Sorry for my delayed reply (I'm a bit behind and only briefly glanced over the diff so far).

> Or better yet, package the LLVM revisions that Chromium distributes that way we don't rely on unverified binaries.

Yes, we should avoid unverified binaries and build as much from source as possible (personally I'm completely against pre-built binaries).

> 1. Should I create a completely new directory in llvmPackages for chromium builds?

Good question :o IIRC Chromium uses LLVM master, I'm not sure how often they update it though.
https://chromium.googlesource.com/chromium/src/+/master/docs/clang.md states:
> It's just upstream clang built at a known-good revision that we bump every two weeks or so.

First, I'd hope they don't update LLVM during patch releases for the stable channel (though IIRC it doesn't take that long to compile, especially vs. Chromium of course). But at least for Chromium dev the LLVM is probably bumped every two weeks and might be enough ahead of the one for the stable channel that it causes problems (i.e. doesn't build). So we could introduce `pkgs/development/compilers/llvm/latest` but `pkgs/development/compilers/llvm/chromium` might be more appropriate.

In any case we should also involve the LLVM maintainers regarding this (IIRC they also considered switching to the monorepo build). Another problem is that the LLVM will probably break after some new updates for the beta/dev channel and we should have a dedicated maintainer to take care of this (I'm already busy enough with Chromium :o).

File: pkgs/applications/networking/browsers/chromium/default.nix
Comment (TredwellGit) [reply pkgs/applications/networking/browsers/chromium/default.nix:33]: Never mind, `llvmPackages.lldClang.stdenv` is totally broken.

**Comment by samuelgrf** [comment]
Closing because I don't have the time to maintain the custom LLVM builds.

**Pull request closed without merging**
